File: api/src/main/java/com/avast/metrics/api/TimerPair.java
package com.avast.metrics.api;

import java.time.Duration;
import java.util.concurrent.Callable;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;

public interface TimerPair {

    interface TimeContext extends AutoCloseable {
        void stop();

        void stopFailure();

        @Override
        default void close() {
            stop();
        }
    }

    TimeContext start();

    void update(Duration duration);

    void updateFailure(Duration duration);

    <T> T time(Callable<T> operation) throws Exception;

    <T> CompletableFuture<T> timeAsync(Callable<CompletableFuture<T>> operation, Executor executor) throws Exception;

}

File: scala-api/src/main/scala/com/avast/metrics/scalaapi/TimerPair.scala
package com.avast.metrics.scalaapi

import java.time.{Duration => JDuration}
import scala.concurrent.duration.Duration
import scala.concurrent.{ExecutionContext, Future}

trait TimerPair {

  trait TimeContext {
    def stop(): Unit
    def stopFailure(): Unit
  }

  def start(): TimeContext

  def update(duration: JDuration): Unit

  def updateFailure(duration: JDuration): Unit

  def update(duration: Duration): Unit

  def updateFailure(duration: Duration): Unit

  def time[A](block: => A): A

  def time[A](future: => Future[A])(implicit ec: ExecutionContext): Future[A]

}
